### ISSUE TITLE
subsequent simultaneous Electron startup very slow

### ISSUE THREAD

**Issue by PerBothner** [electron/electron#13163]
Starting `electron` on a trivial application starts up fine.  However, if I start up `electron` again (while the first command is running) there is a multi-second delay.

* Electron Version: 2.0.2
* Operating System (Platform and Version): Fedora Linux 28 (x86_64)
* Last known working Electron version:  1.8.2

To demonstrate, use this `main.js`:

    const {app,BrowserWindow} = require('electron')
    function createWindow () {
        let openDevTools = true;
        var w = 800, h = 600;
        win = new BrowserWindow({width: w, height: h, show: false})
        win.loadURL('https://google.com')
        if (openDevTools)
            win.webContents.openDevTools()
        win.once('ready-to-show', () => {  win.show()  })
    }
    app.on('ready', createWindow)

and this is the corresponding `package.json`:

    {
      "name"    : "Tester",
      "version" : "1.0.0",
      "main"    : "main.js"
    }

Note the use of `openDevTools` is just to demonstrate the delay. My actual application [DomTerm](https://DomTerm.org) insteads does a bunch of JavaScript, including setting up a WebSockets connection to the `domterm` server.

To demonstrate the testcase:

    $ electron-v2.0.2/electron ~/tmp &
    $ electron-v2.0.2/electron ~/tmp &

The first time the BrowserWindow and the debugger sub-window come up quickly and at the same time.  The second time, the BrowserWindow (including the google.com contents) comes up quickly, but there is a 5-second delay before the debugger sub-window is displayed.

This delay does not happen with electron-v1.8.2.

**Comment by timfish** [comment]
Electron apps should only really be loaded as a single instance. If you want multiple windows you should load them from a single main process.

**Comment by pronebird** [comment]
Try disabling the dev tools, they must be keeping your CPU and disk busy.

**Comment by PerBothner** [comment]
Starting the dev tools is just for the test-case to show that there is a 5-second delay between when the window opens and when the application starts.  I get the same delay in my actual application, which does not normally start the dev tools.  Also note the delay is only on *subsequent* startup of the application, not the first time.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v4.2.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons. This issue will be closed 7 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by PerBothner** [comment]
Still seeing the same problem on v6.0.0-beta.8 and v8.0.0-nightly.20190807.

**Comment by cdetamble** [comment]
> 
> 
> Starting the dev tools is just for the test-case to show that there is a 5-second delay between when the window opens and when the application starts. I get the same delay in my actual application, which does not normally start the dev tools. Also note the delay is only on _subsequent_ startup of the application, not the first time.

Can confirm that this is still the case as of v7.1.7 under Windows 10 (Build 16299). The first instance is not affected whereas subsequent ones are. Looks like the subsequent instances try to connect to a port that is blocked by the first instance and run into a timeout or something.

Anyone knows a solution to this yet? Here's what I already tried:

- Starting subsequent instances from a copied Electron folder.
- Disabling all network adapters.
- Disabling "Auto-Detect Proxy Settings" in Windows' Internet Options.
- Getting rid of all JS code in the `index.html` that is loaded on startup.
- Debug the main process with `inspect-brk`.

**Comment by cdetamble** [comment]
I finally got it to work using `singleInstanceLock`. The idea is that only a single Electron instance exists at any time and subsequent instances arrive in the primary's `main.js` as well. There, I simply open a new window for the subsequent instance.

See https://electronjs.org/docs/api/app#apprequestsingleinstancelock

**Comment by PerBothner** [comment]
Thanks, @cdetamble !  I fixed DomTerm's `main.js` based on this clue and it works great.  (I haven't disabled the server-side work-around I implemented, mainly because it is still needed for the Qt front-end, though I have old logic that could be revived.)

**Comment by MJomaa** [comment]
For everyone who still has problems, those two things helped me:

1. Make sure not to use localStorage at startup. The lock on the localStorage object will eventually time out. This was true for version 7.xx, not sure if this is still true for 8.xx. You can create a settings file at 'userData' instead (and also watch the file if you want to sync settings during run-time).

2.  Have only one app active at one time.  For this you need to save all window instances in one app and close subsequent app launches. 

Save instances globally so they are not garbage colllected.
```
const instances: BrowserWindow[] = [];
```

Don't forget to add the window to the instances collection.
```
const createMainWindow = (args?: string[]) => {
    ...
    instances.push(mainWindow);
};
```
So now here is the magic. An app can have multiple instances/windows. But we only need one app. The `second-instance` event will take care of adding the new instance/window. 

```
// Bootstrap App
const gotTheLock = app.requestSingleInstanceLock();
if (!gotTheLock) {
  app.quit();
} else {
  app.on('second-instance', (event, commandLine, workingDirectory) => createMainWindow());
  app.on('ready', () => createMainWindow());
  if (process.platform === 'darwin') {
    // Quit from the dock context menu should quit the application directly
    app.on('before-quit', () => {
        cmdQPressed = true;
    });
  }
  app.on('window-all-closed', () => {
    if (cmdQPressed || process.platform !== 'darwin') {
      app.quit();
    }
  });
  app.on('activate', () => createMainWindow());
}
```

If you use splash screen instances as entry points it only gets slightly more complicated. You need to generate a session id to associate a splash screen instance with a main window instance and save them in hash maps instead of an array. Hit me up if you do.

**Comment by sunjinbo321** [comment]
> 启动`electron`一个简单的应用程序时一切正常。但是，如果我再次启动`electron`（当第一个命令正在运行时），就会出现几秒钟的延迟。
> 
> * 电子版本：2.0.2
> * 操作系统（平台和版本）：Fedora Linux 28 (x86_64)
> * 最新已知工作 Electron 版本：1.8.2
> 
> 为了演示，请使用以下命令`main.js`：
> 
> ```
> const {app,BrowserWindow} = require('electron')
> function createWindow () {
>     let openDevTools = true;
>     var w = 800, h = 600;
>     win = new BrowserWindow({width: w, height: h, show: false})
>     win.loadURL('https://google.com')
>     if (openDevTools)
>         win.webContents.openDevTools()
>     win.once('ready-to-show', () => {  win.show()  })
> }
> app.on('ready', createWindow)
> ```
> 
> 对应的是`package.json`：
> 
> ```
> {
>   "name"    : "Tester",
>   "version" : "1.0.0",
>   "main"    : "main.js"
> }
> ```
> 
> 请注意，使用`openDevTools`只是为了演示延迟。我的实际应用程序[DomTerm](https://DomTerm.org)执行了一堆 JavaScript，包括设置与`domterm`服务器的 WebSockets 连接。
> 
> 演示测试用例：
> 
> ```
> $ electron-v2.0.2/electron ~/tmp &
> $ electron-v2.0.2/electron ~/tmp &
> ```
> 
> 第一次打开 BrowserWindow 和调试器子窗口时，它们会同时快速出现。第二次打开 BrowserWindow（包括 google.com 内容）时，它们会快速出现，但调试器子窗口会延迟 5 秒才显示出来。
> 
> 使用 electron-v1.8.2 时不会发生这种延迟。

async function createWindow() {
const config = {
...
webPreferences: {
...
}
}
const gotTheLock = app.requestSingleInstanceLock()
if (!gotTheLock) {
config.webPreferences['partition'] = persist_${process.pid}
}

We can set the partition attribute to isolate sessions, which can solve the problem of slow startup and ensure that the first Electron instance can access the LocalStorage cache, but the second instance or more still cannot be resolved

**Comment by sunjinbo321** [comment]
> 启动开发工具只是为了测试用例，以显示窗口打开和应用程序启动之间有 5 秒的延迟。我在实际应用程序中也遇到了同样的延迟，它通常不会启动开发工具。另请注意，延迟仅在应用程序的_后续_启动时发生，而不是第一次启动时。

async function createWindow() {
const config = {
...
webPreferences: {
...
}
}
const gotTheLock = app.requestSingleInstanceLock()
if (!gotTheLock) {
config.webPreferences['partition'] = persist_${process.pid}
}

We can set the partition attribute to isolate sessions, which can solve the problem of slow startup and ensure that the first Electron instance can access the LocalStorage cache, but the second instance or more still cannot be resolved
